TRD: DCS Chamber Status logic for processing

leo-barreto opened the pull request:
## Main goal
Add the processing logic of TRD chamber status and config tags data points.

The proposed procedure is: after a new run begins, first instances of `trd_fedChamberStatus` and `trd_fedCFGtag` get written to CCDB with run number as metadata. Following changes should be logged as alarm (first change), then warnings.

## Minor modifications
- enabled processing and upload to CCDB of `trd_cavernHumidity` and `trd_fedEnvTemp`
- fixed issues with `trd_runNo` and `trd_runType` processing.
- altered macro to read new CCDB objects.
- changed some DCS simulation points for better testing.

davidrohr committed:
GPU: Use double-pipeline within DPL

mwinn2 committed:
added getPitch for DEDigitizer calc.

average position

sync with mch_hit_improvements_4

fix cherrypick

restore header

Clang

clang

clang

fixes Philippe

typo

mfasDa committed:
[EMCAL-918] Handling corrupted trailer words

Corrupted trailer words are words with the trailer
marker but an invalid trailer word code, for which
the trailer word type cannot be determined. Handled
in the RawReaderMemory as MinorRawDataError (non-
crashing)

njacazio committed:
Add labelled string array

matthias-kleiner committed:
Debug streamer: add sampling by weight

- Storing weights and sampling frequency in debug tree

robincaron13 committed:
LTF and CA tracking parameters refined (alignment considerations)

noferini committed:
adjusting TOF calib info alignment to BC

lietava committed:
dev: get lhc filling from LHC IF

lietava committed:
clang

sawenzel committed:
MCTrack: Generalize hit bit property

So far, every possible sensitive detector for ALICE Run3
had a dedicated bit assigned inside MCTracks. The number of such
bit was fixed to be 22 maximally.
In this way we could store information if a particular Monte Carlo track
caused energy deposits in detectors.

Since more and more detectors are being added to the code
(ALICE3, ITS3, R&D), we'll soon have more than 22 DetIDs in our code
which will conflict with the space available in MCTrack.

Since enlarging the MCTrack class in memory is to be avoided, this commit
proposes a solution based on an intermediate lookup table which dynamically
maps the detector ID to a particular bit. The lookup table is setup during
geometry construction and stored in MCEventHeaders for offline use.

This solution should scale much further than before. The limit is now 22 simultaneous
sensitive detectors in the geometry and not 22 total sensitive detectors in the code.

aphecetche committed:
MCH: fix bad-channels-ccdb metadata default type

and add Created metadata tag for default object.

davidrohr committed:
GPU Workflow: Workaround for multi-threaded pipeline step 1: alternate between pipeline contexts

davidrohr committed:
GPU TPC: Send TPC trigger words with orbit information

Commit:
Removing overlaps of carbon plates

davidrohr committed:
GPU TPC: Always send TPC TRIGGERWORD in gpu-reconstruction, sending empty dummy word when processing data that has no triggers

davidrohr committed:
GPU TPC: always enable triggerHandling in tpc-reco workflow and in gpu-reco workflow if caclusterer is active

davidrohr committed:
Fix missing include

shahor02 committed:
Add TPC trigger info to CTF

shahor02 committed:
Add TPC trigger reader/writer

The TPC triggers will be automatically read and written at the same time when
native clusters are read or written, although for technical reasons they are
stored in separate tree/file via dedicated writer.
In case previously produced data set lacking TPC triggers output file needs
to be read-out, one should export DPL_DISABLE_TPC_TRIGGER_READER=1 env.var.
to not initialize the trigger reader.
Triggers are added to the RecoContainer and can be accessed via
recoContainer.getTPCTriggers()

ktf committed:
DPL: inject missing data (#11925)

ktf committed:
DPL: cleanup DataAllocator

* Remove references to the now unsupported boost serialization
* Unify common codepaths

martenole committed:
Fix TRD pile-up information (#11924)

* Fix TRD pile-up and PH information

- pile-up setting was OK, but getters were broken
- after TRD digits were sorted by default the neighbouring check was broken
- due to shared digits PH values could be counted multiple times and incomplete
- fix ClassDef version of CalT0 class

* Fix digit comparison independently of sorting order

frmanso committed:
Update HeatExchanger.cxx

Modified headers

martenole committed:
Fix disable output file writing of ED in FST

shahor02 committed:
gpu-reco request tpc-trigger outputs only with raw input

wiechula committed:
TPC: Add DCAr histograms to TPC QC Tracks component (#11909)

* [WIP] Adds: O2 DCA histograms in tracks task

* Fix DCA histogramming part + cleanup

* Configuration of Propagator must be done outside the code (e.g. QC
  framework)
* map must use std::string as key, instead of std::string_view
* remove unnecessary DCAz
* simplify code
* remove unnecessary includes

---------

Co-authored-by: Bhawani Singh <bhsingh@cern.ch>

sawenzel committed:
o2-sim: Better VMC shutdown (via plugin code)

Introducing the ability to close/terminate the
VMC simulation engines.

* for instance to print final statistics about
  steps, magnetic field calls etc.

mfasDa committed:
[EMCAL-795] Reject ALTRO bunches with impossible length and pages with impossible FEC ordering

davidrohr committed:
DPL: Shorten and downscale message when injecting missing data

davidrohr committed:
Sync workflow: inject missing data at raw-proxy level, and don't expect optional data from TPC

davidrohr committed:
GPU: Clarify some function names

ktf committed:
Bump catch2 to 3.4.0 to try C++20 support (#11930)

davidrohr committed:
GPU TPC: Add option to downscale TPC dEdx computation

ktf committed:
DPL: do not inject DISTSUBTIMEFRAME

davidrohr committed:
GPU Standalone Benchmark: fix enabling dEdx

shahor02 committed:
Add CTP Lumi inputs to its CTF header (#11935)

Co-authored-by: shahoian <ruben.shahoyan@cern.ch>

martenole committed:
TRD raw stats add TFID info (#11888)

davidrohr committed:
DPL / TPC: Fix printing 0xDEADBEEF alarm when using RawPageSequencer

davidrohr committed:
GPU: Fix initialization of the 2 GPUReconstruction instances for double-pipeline mode

shahor02 committed:
Extract both lumis from CTP CTF

Also the HBF count for the 2nd lumi was not propagated to the CTF,
now it is done (though it is anyway the same as the main lumi HBF counts)

davidrohr committed:
GPU: Add gpu-reconstruction-prepare device, forwarding the pointers to gpu-reconstruction

shahor02 committed:
Add options to rescale CTP lumi and select its source used for TPC corrections

davidrohr committed:
GPU: Implement out-of-band message from prepare device to GPU device transfering ptrs to input data

shahor02 committed:
Fix TPC CTF in the linkdef

davidrohr committed:
GPU: Make receiving from out-of-band channel asynchronous

davidrohr committed:
GPU Standalone multi-threading: handle end-of-stream over out-of-band channel

davidrohr committed:
TPC: Support callback in SectorCompletionPolicy that defines an ordering

davidrohr committed:
GPU: Ensure that gpu-prepare-workflow and gpu-workflow process time frames in the same order via completion policy

davidrohr committed:
GPU Workflow: Run 2 threads for the async pipelines, and dispatch the jobs via them

davidrohr committed:
GPU: Make setting updated calib objects / applying them thread-safe

davidrohr committed:
GPU: Combine consecutive calib updates if not applied meanwhile

davidrohr committed:
GPU TPC: Cache TPC triggers internally for some time, and output them only when outputting clusters, allowing for late receival of outputRegions in case of multi-threaded pipeline

davidrohr committed:
GPU: Add optional callback to wait for update of inputs (and outputs) before clusterization output

davidrohr committed:
GPU Workflow: start processing early asynchronously if possible, update inputs and outputcallback when DPL context of TF is available, wait for end of async processing, and ship data out via DPL

lietava committed:
fix: in CTPsaclers det counts are replaced by input counts needed for lumi (det counts were not used)

lietava committed:
clang

davidrohr committed:
GPU: Finalize double-pipeline processing earlier, to prevent deadlock when waiting for late inputs

davidrohr committed:
GPU Interface: Support a callback for providing final set of inputs and output buffer late, to be propagated to GPUReconstruciton only after running TPC Clusterization

davidrohr committed:
dpl-workflow: do not event display on MI100 nodes in online operation

davidrohr committed:
GPU Interface: meanwhile we should be able to dump events in double-pipeline mode

davidrohr committed:
GPU Benchmark: Improve error message, since I had forgotten how it works myself...

davidrohr committed:
GPU: Relaxed delayedOutput constraints

shahor02 committed:
Revert TOF CTF precompressed timeTDCInc to uint16_t

To be compatible with existing dictionary

davidrohr committed:
DPL: Rename reworkEnv to reworkTimeslicePlaceholder and apply it also to channel-config

davidrohr committed:
GPU: Support standalone multi-threading with more than 1 GPU per workflow

davidrohr committed:
dpl-workflow: Limit number of nodes in multiplicity computation to 230 to avoid downscaling

davidrohr committed:
GPU: Reduce verbosity

davidrohr committed:
GPU: Improve scheduling of synchronization of finalization of output transfer of previous TF, and notification to foreign chain

shahor02 committed:
Increment ClassDefs

davidrohr committed:
GPU double-pipeline: Do calib updates during last sectors of clusterization, after receiving late input data

davidrohr committed:
GPU: Clean up multi-threading, fix START/STOP/START for the future, don't poll for FMQ states, and better handling to avoid blocking FMQ state changes

martenole committed:
Denote deadbeef error to warning

mfasDa committed:
[EMCAL-1044] Sort towers in DDL according to FEC ID

Channels must be sorted according to FEC ID as the
ALTRO decoder now makes used of channel ordering
to find corrupted channels

davidrohr committed:
GPU QA: Use single GPU QA instance for double-pipeline

davidrohr committed:
FST: Make DD wait time configurable

ktf committed:
DPL: inject firstTForbit and tfCounter when missing (#11953)

* DPL: inject firstTForbit and tfCounter when missing

Use the values from from the first available data header.

* Clarify error messages

---------

Co-authored-by: David Rohr <github@jwdt.org>

davidrohr committed:
GPU SA MultiThreading: transfer SHM region id per pointer not per TF

davidrohr committed:
dpl-workflow: With NUMAGPUIDS, we expect 5 region callbacks not 4

shahor02 committed:
Suppres duplicate matches from ITS track clones on sector edges (#11958)

Co-authored-by: shahoian <ruben.shahoyan@cern.ch>

fmazzasc committed:
Avoid crashes when CTP is not loaded

davidrohr committed:
GPU SA MultiThreading: Run async receiveThread only once DPL invoked run() the first time to make sure everything is initialized

syano0822 committed:
Add the noise map clear function in Detectors/ITSMFT/MFT/condition/src/DCSConfigReader.cxx/.h

ktf committed:
DPL: fix race condition with the delivery of region events (#11961)

* DPL: fix race condition with the delivery of region events

* Update Framework/Core/src/DataProcessingDevice.cxx

davidrohr committed:
DPL: Do not inject at EndOfStream

davidrohr committed:
GPU: Bugfix for GPU QA in double-pipeline mode

aalkin committed:
DPL Analysis: unified getter for a group of simple-typed columns (#11906)

sawenzel committed:
Disable simulation shared mem on APPLE

Does no longer compile with XCode15 and in any case
never worked correctly.

sawenzel committed:
AODProducer: Code cleanup and optimization

* use std::pair for CTP ClassMasks instead of std::vector.
* avoid mem alloc which not necessary
* initialize at declaration and reduce else statements

vkucera committed:
Fix namespace pollution from `std` and `o2::constants` (#11851)

* Remove std namespace from CCDBDownloader.h

* Remove std namespace from DataBlockBase.h

* Remove math namespace from AnalysisDataModel.h

* Clean TPCSimulation/Digitizer.h

* Clean SACDecoder.h

davidrohr committed:
GPU: Fix standalone multithreading, policy order check was not enabled

davidrohr committed:
GPU: Delay only processing of async message not receiving, to fill completionPolicyQueue in time and do not deadlock

shahor02 committed:
Disable strangeness tracking in pbpb (on in pp) by default

davidrohr committed:
TPC: Fix SectorCompletionPolicy, should retry instead of wait if all inputs present

noferini committed:
adjust TS size for TOF sync calib

chiarazampolli committed:
Adding plots to ITSTPC matching QC

fgrosa committed:
[ITS3] Implement small fixes for ITS3 geometry

f3sch committed:
svertexer: bug fix for tpc-only tracks

Code did not get a reference to the object in the trackpool but made local copy,
thus minR was never modified and left at zero which meant that most if not all
tpc-only tracks were rejected in checkV0 due to invaild causality.

Signed-off-by: Felix Schlepper <f3sch.git@outlook.com>

mfasDa committed:
[EMCAL-1045] Skip FMQ payload in case of corrupted RDH

In case RDH header decoding fails the offset to the next
page cannot be reliably determined, consequently the
payload needs to be skipped. Other errors are kept
as minor errors and only the page is skipped.

mfasDa committed:
[EMCAL-534] Remove message about missing trailer word from AltroDecoder

The message is handled in the RawToCellConverter and
pass to the raw error QC, handling in the AltroDecoder
is legacy and became rather confusing to the shift crew.

mfasDa committed:
[EMCAL-834] Discard only events instead of TF in link checker

Only reject those events that have missing links within a
timeframe, and keep events which have all links. For
events rejected add an empty TriggerRecord with
additional trigger bit masking incompletes as rejected.

lietava committed:
ctpdevL input scalers (#11970)

* dev: scaler inputs dinished and tools

* clang

matthias-kleiner committed:
TPC: add option to not sample first n TFs

noferini committed:
fix in TOF matcher for calib

davidrohr committed:
GPU Workflow: Fix refactoring of input processing

mfasDa committed:
[EMCAL-1046] Swap HG and LG at overflow cut

The reconstruction swap HG and LG at 950 ADC counts
in order to avoid saturation of the HG (i.e. due to
pedestals). The compression instead used the 10 bit
ADC range (1024). In the encoding LG cells in that
energy range are encoded with negative values and
therefore are out-of-range. In addition introduce
a 200 MeV safety margin for the LG in order to not
truncate for cells which are slightly below the
saturation offset due to floating point imprecision.

shahor02 committed:
Stop CTFreader asap to avoid empty run cycle

shahor02 committed:
Fix CTP entropy encoder to handle dummy lumi inputs

shahor02 committed:
Protect CorrectionMapsLoader against dummy LumiInfo

shahor02 committed:
Protect RecoContainer against empty LumiInfo input

davidrohr committed:
GPU: Clean up deletion of old calib objects

fmazzasc committed:
Strangeness tracking: tighter cut on grid granularity (#11979)

* Strangeness tracking: tighter cut on grid granularity

* Please consider the following formatting changes

---------

Co-authored-by: ALICE Action Bot <alibuild@cern.ch>

davidrohr committed:
GPU: Fix merging of calib updates

freidt committed:
ALPIDE event decoding: added  FLUSHED_INCOMPLETE and STROBE_EXTENDED (#11988)

* ALPIDE event decoding: added errors for FLUSHED_INCOMPLETE and STROBE_EXTENDED

* Please consider the following formatting changes

---------

Co-authored-by: ALICE Action Bot <alibuild@cern.ch>

shahor02 committed:
add tpc-triggers to tpc-reco when root output is requested

shahor02 committed:
TrackStudy should process only selected sources

davidrohr committed:
dpl-workflow: make GPU double-pipeline the default for EPN and FST

matthias-kleiner committed:
TPC: adding material type for dedx calibration as parameter

shahor02 committed:
Check if FileFetcher is done for early stop on empty queue (fix for PR11985)

fgrosa committed:
[ITS3] Improve transformation of coordinates and digitisation (#11965)

* [ITS3] Improve transformation of coordinates and digitisation

* Please consider the following formatting changes

---------

Co-authored-by: ALICE Action Bot <alibuild@cern.ch>

jmunozme committed:
Add Alice3 absorber

gvolpe79 committed:
Adding aluminium absorbers in front to chambers 2 and 4

shahor02 committed:
Fix: CTPinputs from prev BC were filled for BCs w/o CTP input

MichaelLettrich committed:
[rANS] Enforce Consistent Histograms

MichaelLettrich committed:
[rANS] Enforce Consistent Histograms

lietava committed:
ctpinputs-decoding-ctf to ignore-ctpinputs-decoding-ctf

martenole committed:
Make TRD raw reader more fault tolerant, allow fixing n time bins

shahor02 committed:
Transfer all lumi scaling settings between alternating CorrectionMapsLoaders

noferini committed:
check fill scheme to clean FT0 BC in TOF calibs

martenole committed:
TRD PH calib send dummy message even if STOP requested

ktf committed:
Avoid injecting data unexpectedly (#11995)

This protects the case of multiple pipelines.

ktf committed:
DPL: improve message for missing data

ktf committed:
Fix macro tests when using boost 1.82 / 1.83 (#12007)

mfasDa committed:
[EMCAL-1048] Treat faulty memory size reported by RDH

In case the memory size in the RDH reported is larger
then the raw payload size forward the current positon
to payload size or offsetToNext - whatever is smaller,
before raising the error.

In case the trailer is incomplete the page cannot be trusted,
likely because it is overwritten in memory. The marker must
be forwarded and an exception is thrown.

MichaelLettrich committed:
[rANS] Fix Compiler Warnings

MichaelLettrich committed:
[rANS] Fix Compiler Warnings

shahor02 committed:
Decrease max error count, err->warn for *Wrong order/duplication*

ktf committed:
Fix deprecation warning (#12006)

davidrohr committed:
MID: Downscale InfoLogger errors temporarily

ktf committed:
DPL: change ExpirationHandler::Creator to use services

This allows us to easily have some state attached to the creation
of new timeslices and in particular we can then use the DecongestionService
to keep track of the next enumeration to be created. This comes handy
when we have dummy iterations which we want to "rewind" to avoid
messages about empty timeslices.

matthias-kleiner committed:
Enable fetching matLUT only if needed

davidrohr committed:
DPL: Fix typo, should be != (#12017)

jmyrcha committed:
o2-eve: HMPID display, fix in display configuration file read (#11997)

* o2-eve: HMPID display, fix in display configuration file read
o2-eve-workflow: storing HMPID in *.root/*.json file

* clang

* LOG->LOGF

* clang

* geometry for HMP

---------

Co-authored-by: Julian Myrcha <jmyrcha@cern.ch>

iravasen committed:
ITS DCS parser: added configurable url to retrieve RCT headers (#11978)

* added configurable url to retrieve RCT headers

* default ccdb to o2 ccdb internal

noferini committed:
fine tuning TOF calib with FT0

ktf committed:
DPL: drop ReadoutAdapter (#12016)

Pre-DataDistribution prototype not needed at this point.

shahor02 committed:
Fill N nearest neighbours in T and Z for vtx study.

shahor02 committed:
Revert "[rANS] Enforce Consistent Histograms"

This reverts commit d57a2647bff8fc62cf2675f38fbafa066e4f7ede.

shahor02 committed:
Revert "[rANS] Fix Compiler Warnings"

This reverts commit 354fbb1596de09aad1df51c460497d55d1e4ac7f.

Commit:
Dont send CCDB if not calculated

ktf committed:
DPL: protect against multiple additions of 0xccdb queries (#12020)

fmazzasc committed:
[SVertexer] Sort tracks by their decayRef when nthreads>1 (#12003)

* Sort tracks by their decayRef when nthreads>1

* Optimize loops

davidrohr committed:
GPU: Fix compiler warning

mfasDa committed:
[EMCAL-791] Require CCDB access for EMCAL digitizer

EMCAL sim params should be handled via the CCDB,
therefore CCDB access is required for EMCAL.

mfasDa committed:
[EMCAL-791] Require CCDB access for EMCAL by default

Option inverted to disable CCDB access on request.

mfasDa committed:
[EMCAL-1048] Consistent check fo RCU trailer and channel header integrity

- RCU trailer must have a valid trailer marker, valid
  trailer version and the corresponding expected size
  of the trailer
- Channel header must have a valid branch ID (0-1),
  FEC ID (0-9) and ALTRO ID (0,2,3,4)
- Channel headers only accept if bit 30 is set and
  bit 31 is not set

Treating corruptions in case of bitflip of bit 31 leading to
an artificial interpretation of a RCU trailer word in case
the word is the last word on the page, even though the payload
continues on the next page.

f3sch committed:
SVertexer: use mean vertex

Signed-off-by: Felix Schlepper <f3sch.git@outlook.com>

shahor02 committed:
uniform disable-strangeness-tracker opt., enable str.tracker by default

davidrohr committed:
DPL: Fix: move allfound optimization to correct loop

davidrohr committed:
DPL: Add option to raw-proxy to print input sizes per spec

davidrohr committed:
CTF: Print also the total size in size report

davidrohr committed:
dpl-workflow: enable size-report for raw-proxy, and downscale in online mode

davidrohr committed:
DPL: Fix injectMissingData not always marking all inactive pipelines as present when exiting early due to allFound

f3sch committed:
ITS3: Add macros for Cluster Size, Hits and Digits (#11835)

Signed-off-by: Felix Schlepper <felix.schlepper@cern.ch>

mfasDa committed:
[EMCAL-791] Remove attic fields in the SimParams

mTimeDelay, mTimeNoise and the switch to take it
from OCDB 1-1 portings from AliRoot to O2 - no
longer used as the time response is now simulated.

martenole committed:
TRD raw reader update

shahor02 committed:
Add tracks time to trackStudy output

shahor02 committed:
Flag if propation has failed in DCAFitterN

shahor02 committed:
Fix in the Close method

shahor02 committed:
Store TFIDInfo TTree together with the vector

shahor02 committed:
Fix getTime... methodes in the RecoContainer

noferini committed:
TOF: add flag to compressor to skip large payloads (#12040)

chiarazampolli committed:
Fix condition to trigger FATAL in ITSTPC matching QC

shahor02 committed:
Make sure the ROF IR is decoded, otherwise abort link data

ktf committed:
DPL: test matcher behaviour

Since at some point there was a doubt that EMCAL messages where
wrongly matched, here is the unit test to demonstrate myself it's
not actually the case.

iravasen committed:
added a wait time of 30s between consecutive pushes to ccdb

davidrohr committed:
FST: should use new CTF coding scheme

ktf committed:
Rewind input proxy (#12015)

* DPL: change ExpirationHandler::Creator to use services

This allows us to easily have some state attached to the creation
of new timeslices and in particular we can then use the DecongestionService
to keep track of the next enumeration to be created. This comes handy
when we have dummy iterations which we want to "rewind" to avoid
messages about empty timeslices.

* DPL: handle empty input proxy iterations

shahor02 committed:
Warn about CTF/rawTF readers waiting for remote data to be copied

shahor02 committed:
Relax type check for old dictionaries in compat mode, warn only.

Initially some CTF blocks were declared unsigned to store incremenets of ordered data but
in some cases ordering was broken. To not inflate dynamic range due to the neagative -> unsigned
cast we changed these types to signed but in a few old dictionaries we still have data min/max
which does not fit to the signed range. For these old dictionaries we have no other way but to
not throw exception on the failed type check. The (de)coding does not suffer since if done
consistently with the same data type defined, even if it does not match to the data range.

davidrohr committed:
DPL: When enumeration timeframe was rewinded, don't send a newer oldestPossible (#12048)

davidrohr committed:
EMCAL: Downscale raw parser error messages

ktf committed:
Reduce peak memory usage of TrackerTraits::findTracks (#12046)

ktf committed:
DPL: add support for converting various containers to vectors

sawenzel committed:
MatLUT: Option for fast layer lookup (#12009)

* MatLUT: Option for fast layer lookup

Profile analysis has shown that material budget calculations
used in tracking, spent most time in
 (a) determination of a cylindrical layer
 (b) determination of a phi sector in a layer

This commit proposes an improvement for (a). Instead
of determining the layer id via binary search, we can
use a simple lookup structure working on regular R-intervals.

The commit offers a 5-8% speed improvement for ITS reconstruction
and should be beneficial for other things such as strangeness tracking,
or any algorithm using MatLUT.

Relates to https://alice.its.cern.ch/jira/browse/O2-4170

* Make compile on GPU

* Update MatLayerCylSet.cxx

davidrohr committed:
DPL: If in OnlineMode, input proxies should only send EoS if they receive it

davidrohr committed:
DPL: Should not set newRun flag with incoming runNumber 0, as is the case e.g. with EoS messages

matthias-kleiner committed:
TPC: Adding workflow for integrated DCAs for corrections and QC (#11984)

* TPC: Adding workflow for integrated DCAs for corrections and QC

- DCAs are integrated in phi bins over all tgl and in tgl bins
integrated over phi
- DCA integrated per side is also stored
- RMS and number of tracks are stored per bin
- Multithreading for DCA calculation of tracks
- Additional information with per track DCAs can be written with debug
streamer
- Adding start time as meta data to integrated currents
- Adding DCAs for ITS-TPC matched tracks
- Adding collision vertex
- Adding nearest vertex to tracks which have no vertex assigned
- Adding DCAz from time bin delta

Adding ITS-TPC matching efficiency

Adding: sqrt(chi2Match), MIP/dedx, sqrt(chi2), nCl

* Adding time information, fixing bug when creating tree

* Adding ITS-TPC standalone and afterburner efficiencies

* Speed up track propagation

* Fixing range check of bins

martenole committed:
Protection against bogus TRD PH values

shahor02 committed:
remove unnecessary include

shahor02 committed:
Optimize ITS/TPC mathing memory usage

davidrohr committed:
GPU: Minor optimization

Commit:
[MatchITSTPCQC] Skip invalid MC labels

wiechula committed:
Add EOV functionality

noferini committed:
adding paranoid flag in tof-compressor on epn

aalkin committed:
DPL Analysis: correctly account for extended table version with aod-spawner (#12056)

cortesep committed:
ZDC Small fixes and improved printout in waveform calibration (#11900)

* Better printout and fixes

* WIP

* Fixes and cleanup

* Improved raw data debugging

* Reduce output size in waveform calibration

* Update in debug

* Update in debug

* Updated printout

* Fixes

* Please consider the following formatting changes (#82)

* Fix inversion of cuts

---------

Co-authored-by: ALICE Builder <alibuild@users.noreply.github.com>

dstocco committed:
Macro to query or upload MID CCDB objects (#12025)

shahor02 committed:
Add CTP Lumi to rawTF-reader outputs

matthias-kleiner committed:
TPC IDC: Adding check for default GPECS object

martenole committed:
TRD most probably track time is trigger time

matthias-kleiner committed:
Integrated currents: Adding check if any data has been received

noferini committed:
adding protection in tof-compressor

shahor02 committed:
Improve FT0 interaction tag, use for TRD pileup

noferini committed:
fix FEE mismatch in tof-compressor

ktf committed:
Avoid copy of tracks

noferini committed:
add protections in tof-compressor

davidrohr committed:
dpl-workflow: Use new CTF coding by default on EPNs

davidrohr committed:
DPL: If we receive only the EoS for that run, we need to add to the counter after the resetting

martenole reviewed (approved):
approving only to start the CI for first-time contributor to O2

davidrohr committed:
DPL: Check for EoS on all channels before invoking control->endOfStream()

davidrohr committed:
Revert "DPL: If in OnlineMode, input proxies should only send EoS if they receive it"

This reverts commit ed52430476f2bdb566ae5738f62efefb4323df32.

aalkin committed:
DPL Analysis: make sure slice index with start==end is tested

matthias-kleiner committed:
TPC: Using PropagateToXBxByBz for residual gain map extraction

- adjusting default parameters
- small optimizations during normalisation of the extracted map

ktf committed:
DPL: add finaliseOutputs callback for services

This is invoked after the processing is done (therefore no other
outputs are expected from user code) but before the postProcessing
(which is were data is actually sent to their consumers.

ktf committed:
DPL: add injected CCDB messages in finaliseOutputs

This is needed to avoid a race condition where outputs are sent after
the end of stream.

Roberta-Ferioli committed:
ITS-Study: create pt resolution, add efficiency of sec study (#11763)

* create Ptresolution, add efficiency of sec study

* Fix format

* Add Pt res study at TrackChech

* Add Pt resolution at check study

* ITS-Study: add some study on efficiency of secondary tracks, add fake clusters of secondary analysis

* fix formatting

* fix formatting

* ITS-Study: change structure, add Fake cluster study

* ITS-Study:add fake clusters identification

* [WIP] ITS-Study: completed study on fake clusters

* ITS-Study:Completed the study on secondary tracks and fake clusters, start study on Pt resolution

* ITS-Study: completed secondary tracks efficiency study, started studies on fake clusters and Pt resolution

* Update Detectors/ITSMFT/ITS/postprocessing/studies/src/TrackCheck.cxx

* Update Detectors/ITSMFT/ITS/postprocessing/studies/src/TrackCheck.cxx

* Update Detectors/ITSMFT/ITS/postprocessing/studies/src/TrackCheck.cxx

* Update Detectors/ITSMFT/ITS/postprocessing/studies/src/TrackCheck.cxx

* Update Detectors/ITSMFT/ITS/postprocessing/studies/src/TrackCheck.cxx

* Update Detectors/ITSMFT/ITS/postprocessing/studies/src/TrackCheck.cxx

* Update Detectors/ITSMFT/ITS/postprocessing/studies/src/TrackCheck.cxx

---------

Co-authored-by: Matteo Concas <mconcas@cern.ch>

mconcas committed:
GPU: CMake enable_language(HIP) (#11860)

* Enable HIP language and start porting benchmarks

* Use dedicated fucntion to define symbols upon gpuarch

* Remove ROOT dependency from benchmark

* Move hipification into dedicated files

* Cleanup ongoing

* Filter out IGP architectures

* Fix wrong linking options, add arch override

martenole commented on Detectors/TRD/calibration/workflow/TRDDCSDataProcessorSpec.h:
```diff
@@ -314,6 +450,16 @@ DataProcessorSpec getTRDDCSDataProcessorSpec()
+  outputs.emplace_back(ConcreteDataTypeMatcher{o2::calibration::Utils::gDataOriginCDBWrapper, "TRD_DCSFedDPs"});
+  outputs.emplace_back(ConcreteDataTypeMatcher{o2::calibration::Utils::gDataOriginCDBWrapper, "TRD_DCSFedDPs"});
+  outputs.emplace_back(ConcreteDataTypeMatcher{o2::calibration::Utils::gDataOriginCDBPayload, "TRD_DCSFedDPs"});
+  outputs.emplace_back(ConcreteDataTypeMatcher{o2::calibration::Utils::gDataOriginCDBWrapper, "TRD_DCSFedDPs"});
```
I think it might work with having different objects with the same message name, but to be on the safe side I would use unique names. So instead of `TRD_DCSFedDPs` use `TRD_ChamberStatus`, `TRD_CfgTag`, ...

leo-barreto replied:
Done! See commit dd201ca894eeef9e16e4d68763e4f1b74eaf3db5 and commit 139868a927b12865e0b771294cde0c53a3501a58.

martenole commented on Detectors/TRD/calibration/include/TRDCalibration/DCSProcessor.h:
```diff
@@ -80,6 +87,18 @@ class DCSProcessor
   const std::unordered_map<DPID, TRDDCSMinMaxMeanInfo>& getTRDCurrentsDPsInfo() const { return mTRDDCSCurrents; }
   const std::unordered_map<DPID, TRDDCSMinMaxMeanInfo>& getTRDEnvDPsInfo() const { return mTRDDCSEnv; }
   const std::unordered_map<DPID, int>& getTRDRunDPsInfo() const { return mTRDDCSRun; }
+  // LB: new DPs for Fed and Cavern
```
this is really minor, but could you please adapt to the ordering of the class and put all functions returning a `CcdbObjectInfo` in one block and the ones returning the maps below? It just makes it easier to read imo

leo-barreto replied:
Done! See commit a7aba7244ca2b662bad872de36cba2c4cc2bca76.

martenole commented on Detectors/TRD/calibration/include/TRDCalibration/DCSProcessor.h:
```diff
@@ -102,13 +126,20 @@ class DCSProcessor
   // Or send errors to the InfoLogger in case CFG tag mismatches are detected for chamber which have the same FSM state?
   // For this I need more information on the chamber status - which status indicates all good and included in data taking?
   // not TODO
   // I don't think the FED ENV temperature is needed at analysis level at any point in time so I am leaving it out for now
```
I think you can remove this commented TODO block now :)

leo-barreto replied:
Done! See commit a7aba7244ca2b662bad872de36cba2c4cc2bca76.

martenole commented on Detectors/TRD/calibration/include/TRDCalibration/DCSProcessor.h:
```diff
@@ -117,21 +148,46 @@ class DCSProcessor
+  int mCurrentRunNumber;
+  int mFedChamberStatusAlarmCounter{0};
+  int mFedCFGtagAlarmCounter{0};
+  int mFedAlarmCounterMax{1};
```
Could you make this alarm counter configurable from command line? So add a setter in this class and call the setter from `TRDDCSDataProcessorSpec`.

leo-barreto replied:
Done! See commit 0a3cae7a34f6a2890014c67dd8fd92279507820b.

martenole commented on Detectors/TRD/calibration/src/DCSProcessor.cxx:
```diff
@@ -163,42 +165,135 @@ int DCSProcessor::processDP(const DPCOM& dpcom)
+        }
+        auto& dpInfoFedEnvTemp = mTRDDCSFedEnvTemp[dpid];
+        if (etime != mLastDPTimeStamps[dpid]) {
+          // only add data point in case last one was not already read before (every 0.5 hours)
```
but here you are updating the point every time you see it, no? Or where does the 0.5 h come into play?

leo-barreto replied:
The objected is processed everytime, but it only updates the CCDB every 0.5h. This is defined in line 208 of `workflow/TRDDCSDataProcessorSpec.h` as there is a time interval for calling `sendDPsoutputFedEnvTemp()` (see [here](https://github.com/leo-barreto/AliceO2/blob/139868a927b12865e0b771294cde0c53a3501a58/Detectors/TRD/calibration/workflow/TRDDCSDataProcessorSpec.h#L207)). Similar to other already defined objects with time intervals for updating. 

The way this is written makes a change to `TRDDCSMinMaxMeanInfo` type object really straightforward, as suggested in the next [conversation](https://github.com/AliceO2Group/AliceO2/pull/12079#discussion_r1362098858l).

The same thing for `trd_cavernHumidity`.

martenole replied:
ok, but then I would just remove the `(ever 0.5 h)` passage, because if we decide to change it at some point then it will certainly be forgotten to update this comment

leo-barreto replied:
Done! See commit https://github.com/AliceO2Group/AliceO2/pull/12079/commits/98f02ef5f3e9be889e097f12e7eaa0773111f79d.

martenole commented on Detectors/TRD/calibration/src/DCSProcessor.cxx:
```diff
@@ -163,42 +165,135 @@ int DCSProcessor::processDP(const DPCOM& dpcom)
+        auto& dpInfoFedEnvTemp = mTRDDCSFedEnvTemp[dpid];
+        if (etime != mLastDPTimeStamps[dpid]) {
+          // only add data point in case last one was not already read before (every 0.5 hours)
+          dpInfoFedEnvTemp = o2::dcs::getValue<double>(dpcom);
```
if you want to store the FED env temperature (one can do it, but I don't think it would be strictly needed so you could also drop that part) I would also use the `TRDDCSMinMaxMeanInfo` struct so that you don't only have a single value for a calibration interval which might just be an outlier

leo-barreto replied:
Done! See commit https://github.com/AliceO2Group/AliceO2/pull/12079/commits/98f02ef5f3e9be889e097f12e7eaa0773111f79d.

martenole commented on Detectors/TRD/calibration/src/DCSProcessor.cxx:
```diff
@@ -163,42 +165,135 @@ int DCSProcessor::processDP(const DPCOM& dpcom)
+        auto& dpInfoCavern = mTRDDCSCavern[dpid];
+        if (etime != mLastDPTimeStamps[dpid]) {
+          // only add data point in case last one was not already read before (every 2 hours)
+          dpInfoCavern = o2::dcs::getValue<double>(dpcom);
```
same comment as for the FED env temperature

shahor02 committed:
Set Maurice as a codewonwer of MFT

martenole reviewed (changes requested):
Hi @leo-barreto 
thanks a lot for this work. In general it looks already quite good. I have added some minor comments directly in the code.
For the logic detecting run number changes and triggering the upload to CCDB I am not sure it would work the way it is currently implemented. But better to have a quick chat and go though the code together.
Cheers,
Ole

ktf committed:
Fix various codechecker issues

fmazzasc committed:
[Strangeness Tracking] Fix cascade invariant masses (#12091)

* Fix cascade invariant masses

* Please consider the following formatting changes

---------

Co-authored-by: ALICE Action Bot <alibuild@cern.ch>

ktf committed:
Add protection for passing a root file as a list of files

fcatalan92 committed:
Fix typo in subGenerator check

ktf committed:
Make sure we notify DPL about non-DPL message sending

This is needed to avoid skipping CCDB callbacks on the last timeframe.

fmazzasc committed:
Adding cluster size to track extra (#12044)

* Add ITS cluster size to the AO2D

* Add TracksExtra_001

* Update readerhelper

* Add cluster sizes to AB

* Remove its cluster map

* Please consider the following formatting changes

* Add table versioning

* Fix comments

Co-authored-by: Jan Fiete <jgrosseo@cern.ch>

* Fix comments (2)

Co-authored-by: Jan Fiete <jgrosseo@cern.ch>

* Set old TrackExtra as default

* Fix data descriptor for trackextra

* Please consider the following formatting changes

* Fix test

* Clean commented code

* Fix comment (3)

Co-authored-by: Jan Fiete <jgrosseo@cern.ch>

* Move dyn columns to versioning namespace

* Update Framework/Core/include/Framework/AnalysisDataModel.h

* Add braces + unsigned shorts

* Move cluster sizes to uint8_t

* Swap pattern and clsizes order

---------

Co-authored-by: Maximiliano Puccio <maximiliano.puccio@cern.ch>
Co-authored-by: ALICE Action Bot <alibuild@cern.ch>
Co-authored-by: Jan Fiete <jgrosseo@cern.ch>

martenole committed:
Remove obsolete TRD compressor files (#12102)

chiarazampolli committed:
Using http to access CCDB

ktf committed:
Drop obsolete flag

-undefined error is the default now.

davidrohr committed:
DPL: Fix new run detection in input proxy

ktf committed:
DPL: change InjectorFunction to return a bool

The result is now wether or not something was send out of band

ktf committed:
DPL: notify DPL about out of band data being sent by input proxy

This is needed to avoid skipping some of the callbacks, in particular
the CCDB one.

matthias-kleiner committed:
TOF: fetching GRPECS for correct number of orbits per TF

chiarazampolli committed:
Adding also detector list

davidrohr committed:
DPL: Should always check if we sent something before sending CCDB message

davidrohr committed:
DPL: Take runNumber and timeSlice for CCDB message from timingInfo, instead of first random message

amorsch committed:
hooks and configuration for POWHEG

davidrohr committed:
GPU: Fix HIP compilation with modern system LLVM present

ktf committed:
Port the same fix we have in the Run3 code (#12104)

davidrohr committed:
RawFileReader: Fix triggering CCDB message

martenole committed:
TRD Kr cluster writing to EOS (#12113)

mconcas committed:
ITS tracker: Reserve output buffers (#12089)

* ITS tracker: Reserve output buffers

* Rerserve all possible buffers

iravasen committed:
ITS: process hits from different MEBs + bug fix for scurve dump (#12110)

* added possibility to process hits from different MEBs + bug fix for s-curves dump

* reverted unwanted change

sevdokim committed:
CPV,PHS: modernize cell, digit and cluster readers (#12060)

* CPV,PHS: modernize cell, digit and cluster readers

* GlobalTrackingWorkflow: add modernized CPV and PHS readers to InputHelper.cxx

Commit:
Several upgrades for Event-Generators (#12032)

Several improvements for event generators:

- standardization of info keys in MC event headers
- new GeneratorTParticle generator, being able to read simple TParticle events from file or process 
- ability to read HepMC events from process via FIFO and other improvements for HepMC


Summary of individual commit messages 
-----------
* Introduce pre-defined keys for MC event header info

A number of keys into the ME event header information
mapping is defined.   This is to ensure that code will
use the same keys when ever information is set.

Additional, non-predefined keys, are still possible.

This makes it much more robust when we ask for specific
MC information from the event header, such as

- cross-section(s)
- weight(s)
- Heavy-ion "geometry" parameters
  - Npart in projectile and target
  - Ncoll in various views
    - Overall
    - Hard
    - wounded-nucleon on nucleon
    - nucleon on wounded-nucleon
    - wounded on wounded
- Parton distribution function parameters

This is crucial for building a HepMC event structure which
can be passed on to say for example Rivet analyses.

* Export _full_ header information to MC event header

The generator has been changed so that it exports
_all_ relevant and available information from Pythia to
the MC event header, including heavy-ion "geometry"
parameters.  In particular, the information is stored
in an HepMC compatible way for later user by f.ex.
Rivet.

Note, the current code counts up the number of collisions
by it self.  However, the authors of Pythia have another
way of doing that.   The code is now there to do it the
same way as the Pythia authors, but is currenly disabled.

We should decide which is the appropriate way to count
Ncoll.  I would recommend to follow how the Pythia
authors do it.

* Full header read-in and external program

This change does two things:

**Full header**

_All_ information available in the HepMC event header is
propagated to the MC event header information map.  This
includes

- Heavy-ion "geometry" parameters (b,Ncoll,Npart,...)
- Cross-section(s)
- Weight(s)
- PDF information
- and other attributes defined

This is so that we can build a full HepMC event structure later -
for example to pass to Rivet analyses

** External program **

The functionality of the generator is expanded so that it may
spawn an event generator program, say `eg`.

- The generator opens a FIFO
- The generator then executes the program `eg` in the background
  - The `eg` program is assumed to write HepMC event records on
    standard output, which is then redirected to the FIFO
- The generator reads events from the FIFO

For this to work, a number of conditions _must_ be met by the
`eg` program:

- It _must_ write events in the HepMC event format
- It _must_ write the HepMC event records to standard output
- It _cannot_ write anything else but the HepMC event record to
  standard output
- It _must_ accept the command line option `-n NEVENTS` to
  set the number of events to generate.

If a particular `eg` program does not meet these requirements, then
a simple shell script can be defined to wrap the `eg` appropriately.
For example, the CRMC program `crmc` _can_ write HepMC events to
standard output, but it will also dump other stuff there.  Thus,
we can provide the script

    #!/bin/sh

    crmc $@ -o hepmc3 -f /dev/stdout | \
       sed -n 's/^\(HepMC::\|[EAUWVP] \)/\1/p'

which simply filters the output of `crmc`.  Another EG program
may not accept the `-n EVENTS` command line option, but rather has
the command line option `--nevents`, so then we would do something
like

    #!/bin/sh
    cmdline="eg-program -o /dev/stdout "

    while test $# -gt 0 ; do
       case x$1 in
       x-n) cmdline="$cmdline -n $2"; shift ;;
       *)   cmdline="$cmdline $1" ;;
       esac

       shift
    done

    $cmdline

The command line to run is specified as

    --configKeyValues "HepMC.progCmd=<program and options>"

and can include not only the program name but also other
options to the program.  For example

    --configKeyValues "HepMC.progCmd=crmc -m 5 -i 20800820 -I 20800820"

for Pb-Pb collisions with Hijing.

With this change, we can use _any_ event generator which is capable to
write out its event records in the HepMC format.

* New generator GeneratorTParticle

The generator `GeneratorTParticle` will read in particles
from a `TChain` containing a branch with a `TClonesArray` of
`TParticle` objects.

The generator can operate in two modes

- Data is read from a file(s)
- Data is read from a file being generated by a child
  program

The first mode is selected by

    -g tparticle --configKeyValues "TParticle.fileNames=foo.root,bar.root"

The second mode is selected by

    -g tparticle --configKeyValues "TParticle.progCmd=<program and options>"

For this latter mode, see also recent commit to `GeneratorHepMC`

Above `<program and options>` specifiy a program to spawn in the
background which will write to a specified file (temporary file).
Suppose the program is called `eg`, then the following _must_ be
possible

    eg -n NEVENTS -o OUTPUT_FILENAME

That is, `eg` _must_ accept the option `-n` to set the number of
events to produce, and the option `-o` to set the output file name
(a ROOT file).

The name of the `TTree` object in the file(s) can be set with

    --configKeyValues "TParticle.treeName=<name>"

(defaults to `T`), and similar for the branch that contains the
`TClonesArray` of `TParticle`

    --configKeyValues "TParticle.branchName=<name>"

(defaults to `Particles`).

The generator `GeneratorTParticle` _does not_ import any header
information into the simulation event record.   Some proper
convention could be decided upon, e.g., one that tracks the
HepMC event record format.

* Refactoring to HepMC and TParticle Generators

The classes `GeneratorHepMC` and `GeneratorTParticle` is refactored to
derive from the (second) base class `GeneratorFileOrCmd`.

`GeneratorFileOrCmd` provides common infrastructure to specify

- File(s) to read events from (ROOT files in case of
  `GeneratorTParticle`, and HepMC files in case of `GeneratorHepMC`),
  _or_
- Which commmand to execute and with which options.

It also provides infrastructure to make unique temporary names, a FIFO
to read from (child program writes to), and so on.

These are all configured through configuration keys prefixed by
`GeneratorFileOrCmd.`.

Other changes include

- `GeneratorHepMC` will open _any_ file that HepMC supports - ASCII,
  compressed ASCII, HEPEVT, etc.
- Through the use of `GeneratorFileOrCmd` the command line option flags
  for specifying seed (default: `-s`), number of events (default `-n`),
  largest impact parameter (defautl: `-b`), output (default: `>`), and
  so on can be configured via configuration key values
- `GeneratorHepMC` and `GeneratorTParticle` are passed the
  `GeneratorFileOrCmdParam` as well as specific `GeneratorHepMCParam`
  and `GeneratorTParticleParam`, respectively, objects by
  `GeneratorFactor` and sets the internal parameters accordingly. This
  hides the specifics of the parameters from `GeneratorFactory`.

- `GeneratorHepMC` accepts the old `HepMC.fileName` configuration
  option (with a deprecation warning)

- `HepMC.version` is honoured when spawning a child process.
  That is, if it is set to 2, then use compatibility reader
  `HepMC3::ReaderAsciiVersion2`, otherwise use the version 3
  reader.

  If option is given with file names, then give warning that
  it is not used.

shahor02 committed:
Add methods to get beam energy in GeV, make getSqrtS to return GeV/nucleon pair

sawenzel committed:
MatLUT: faster phi-sector determination; reduction of divisions (#12030)

Following ticket https://alice.its.cern.ch/jira/browse/O2-4170,

Profile analysis has shown that material budget calculations used in tracking, spent most time in
(a) determination of a cylindrical layer
(b) determination of a phi sector in a layer

This commit proposes an improvement for (b):

- Instead of using system function phi = atan2(y, x), we use a
  faster - approximate version. We can do this, since we never need
  to use the precise result of phi in a calculation. Instead, phi is merely used
  to lookup a slice/sector.
  The error of the approximate version is ~1e-5, which is enough to get the correct phi
  sector in 99.999 of the cases.

  (Rare cases, where we end up in a wrong neighbouring phi sector are probably ok ... since
   the material budget lookup is anyways an approximation.)

In addition, we propose a smaller code mofication for ray.CrossZ to reduce
the number of divisions and if statements.
A previous check for
ray orthogonal to z can not happen in the code since there is a previous condition
`if (zID != zIDLast)`.

ITS reco is shown to improve by ~15% in speed for larger PbPb MC timeframes
(on Linux - tested on CC7 and Ubuntu22.04)

matthias-kleiner committed:
Debug streamer: Add downsampling by tsalis (#12112)

* Debug streamer: Add downsampling by tsalis

* Adding time series for vertex position

- adding time series for DCA to vertex for ITS-TPC tracks

* increase ClassDef for TimeSeriesITSTPC

* fix typo

mpuccio committed:
Flatten cell neighbours structure (#12121)

shahor02 committed:
Protect MatLut against dZ=0 (#12124)

Co-authored-by: shahoian <ruben.shahoyan@cern.ch>

mconcas committed:
Update CODEOWNERS for upgrades (#12125)

ddobrigk committed:
Revert "Protect MatLut against dZ=0 (#12124)"

This reverts commit 04ce1274f237a6fd89368ac00134a0a408bed39a.

shahor02 committed:
Revert "MatLUT: faster phi-sector determination; reduction of divisions (#12030)"

This reverts commit 030e39648d85863cc4d89e57df98498c4f53cfc1.

alibuild commented:
Error while checking build/O2/fullCI for 92b91dee5031e574256466f2d40df3a0ec1409bc at 2023-10-26 06:33:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/QualityControl-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.
[ERROR] Function RecoDecay::getMassPDG is deprecated and will be removed soon.
[ERROR] Please use the Mass function in the O2DatabasePDG service instead.
[ERROR] See the example of usage in Tutorials/src/usingPDGService.cxx.


## sw/BUILD/O2-full-system-test-latest/log
Detected critical problem in logfile reco_NOGPU.log
reco_NOGPU.log:[93543:calib-emcalchannel-badchannel]: [04:33:13][FATAL] error while setting up workflow in o2-calibration-emcal-channel-calib-workflow: filesystem error: cannot create directories: Permission denied [/scratch/services/detector_tmp/emc_calib]
[93539:BadMapCalibSpec]: [04:32:42][ERROR] Insufficient statistics: 7 entries in lowE histo, do nothing
[93696:ctf-writer]: [04:33:13][ERROR] Some Lifetime::Timeframe data got dropped starting at 1
[93543:calib-emcalchannel-badchannel]: [04:33:13][FATAL] error while setting up workflow in o2-calibration-emcal-channel-calib-workflow: filesystem error: cannot create directories: Permission denied [/scratch/services/detector_tmp/emc_calib]
[ERROR] pid 93543 (calib-emcalchannel-badchannel) crashed with or was killed with exit code 1


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
[0 more errors; see full log]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12079/92b91dee5031e574256466f2d40df3a0ec1409bc/build_O2_fullCI/pretty.html).

leo-barreto commented:
The new logic for updating the ChamberStatus and CFGtag objects is based on two conditions:

1. Check if at least minimum number `mFedMinimunDPsForUpdate`  of unique id ChamberStatus/CFGtag DPs are sent for processing. The min value is a configurable, with default value of 540, in case some Chamber is known to not send data, for example. The uniqueness check is done using vectors sizes.

2. Check if a new run begins given a change in `trd_runNo` object when compared to previous one, only considers when new entry is positve (as expected from a run number).

If both conditions are true, ChamberStatus/CFGtag DPs are uploaded to CCDB.
If at least one condition is false but there is a change in the DPs, the original alarm/warning method for logging kicks in.

Edit: In commit https://github.com/AliceO2Group/AliceO2/pull/12079/commits/b2ae9dbe7f57f9bd97302ac53b1109633a2214b7, conditions for run updated are split between ChamberStatus and CFGtag, each one resets after the update of its respective object. Resetting after Run update would imply in missing possible updates if two instances of `trd_runNo` were sent before ChamberStatus/CFGtag in new run

martenole commented on Detectors/TRD/calibration/include/TRDCalibration/DCSProcessor.h:
```diff
@@ -117,24 +139,52 @@ class DCSProcessor
+  bool mFedChamberStatusCompleteDPs{false};
+  bool mFedCFGtagCompleteDPs{false};
+  bool mNewRunForFedChamberStatusUpdate{false};
+  bool mNewRunForFedCFGtagUpdate{false};
```
why do you need to flags here? Shouldn't there only be one 'mNewRun' needed?

leo-barreto replied:
Alright, this is connected to the other conversation and I'll answer both here:

We need 4 variables in total, two for each ChamberStatus and CFGtag. This is due to the fact that the objects get updated to CCDB separately and the flags are cleaned after the update:

- The flags considering if the DPs are complete must be two (for each), because one object can be incomplete but not the other (at least in simulated data).
- The "run flags" need to also be separated because once one object is updated, it would "block" the updated of the following. For example, if the flag `mNewRun` was used for both: a new run begins -> `mNewRun = true`, ChamberStatus is updated -> `mNewRun = false` (to just upload first entry) -> CFGtag cannot be updated.

This flag also cannot be reset after a run update in case a new run number starts -> `mNewRun = true`, ChamberStatus is incomplete, run is updated -> `mNewRun = false`, complete ChamberStatus is sent but now it can't be updated.

I changed the name of variables to make this clearer, see commit https://github.com/AliceO2Group/AliceO2/pull/12079/commits/c1e9bcaa8af2752225fb1fd1165e3715813d0243.

martenole commented on Detectors/TRD/calibration/src/DCSProcessor.cxx:
```diff
@@ -59,6 +59,43 @@ int DCSProcessor::process(const gsl::span<const DPCOM> dps)
+  // LB: check if all ChamberStatus/CFGtag DPs were sent by using a vector and then
+  // checking for duplicates, if new size is equal to 540 => all DPs were sent
+  std::vector<DPID> allChamberStatusDPsid;
+  std::vector<DPID> allCFGtagDPsid;
```
I think per iteration a DP cannot be send more than once. So the `std::unique` below should not be needed. To be sure, you could just use a `std::set` here which in any case only holds a unique set of objects.
But I would rather use the `mPids` map which is already there. Can you not check this if a sufficient number of DPs of a certain type have been seen?

leo-barreto replied:
I also think the DP cannot be duplicated, so I removed this part of the code and changed for a simple counter given the entries of `dps`. Since assuming duplicates adds some complexity/confusion to the code, I think it is better to go without it and check the logs if there are duplicates, then we can do a quick fix.

We cannot use the `mPids` to track if ChamberStatus/CFGtag are complete, because the map is filled with the configuration information. We also cannot track if all objects following a pattern were changed to true (after processing) because we need this information before calling the processor. Maybe you meant `mapin`? I changed the code so the check of pattern is done in the same loop as `mapin`.

See commit https://github.com/AliceO2Group/AliceO2/pull/12079/commits/c1e9bcaa8af2752225fb1fd1165e3715813d0243.

martenole commented on Detectors/TRD/calibration/src/DCSProcessor.cxx:
```diff
@@ -163,42 +202,116 @@ int DCSProcessor::processDP(const DPCOM& dpcom)
+          if (o2::dcs::getValue<int32_t>(dpcom) != mCurrentRunNumber) {
+            LOG(info) << "New run number " << o2::dcs::getValue<int32_t>(dpcom) << " differs from the old one " << mCurrentRunNumber;
+            mShouldUpdateRun = true;
+            // LB: two different flags as they reset separately, after upload of CCDB, for each object
```
why would you want to use two separate variables?

leo-barreto replied:
See https://github.com/AliceO2Group/AliceO2/pull/12079#discussion_r1379732422.

martenole commented on Detectors/TRD/calibration/workflow/TRDDCSDataProcessorSpec.h:
```diff
@@ -116,16 +128,40 @@ class TRDDCSDataProcessor : public o2::framework::Task
+
+    // LB: set maximum number of alarms in change in FedChamberStatus and FedCFGtag
+    int alarmfed = ic.options().get<int>("DPs-max-counter-alarm-fed");
+    if (alarmfed > 0) {
```
I would also allow zero so that no alarm is printed

leo-barreto replied:
Done! See commit https://github.com/AliceO2Group/AliceO2/pull/12079/commits/d5011ff8800ba4fa5d17263e1931ef2d57f3b0bc.

martenole reviewed (commented):
Just a few more comments from my side. Then, I think we can merge and test it in the real system

alibuild commented:
Error while checking build/O2/fullCI for c1e9bcaa8af2752225fb1fd1165e3715813d0243 at 2023-11-08 16:46:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2Physics-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/QualityControl-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/O2-full-system-test-latest/log
Detected critical problem in logfile digi.log
digi.log-[31139:internal-dpl-ccdb-backend]: [15:45:50][FATAL] Unable to find object CTP/Calib/OrbitReset/1550600800000
digi.log:[31139:internal-dpl-ccdb-backend]: [15:45:50][ERROR] Exception caught: Fatal error 
[31139:internal-dpl-ccdb-backend]: [15:45:50][ERROR] Failed to open file alien:///alice/data/CCDB/CTP/Calib/OrbitReset/01/35179/b07c53e0-b4c0-11ec-b66d-90ce809b250c?filetype=raw
[31139:internal-dpl-ccdb-backend]: [15:45:50][ERROR] Did not receive content for CTP/Calib/OrbitReset
[31139:internal-dpl-ccdb-backend]: [15:45:50][FATAL] Unable to find object CTP/Calib/OrbitReset/1550600800000
[31139:internal-dpl-ccdb-backend]: [15:45:50][ERROR] Exception caught: Fatal error 
[31139:internal-dpl-ccdb-backend]: [15:45:51][FATAL] Unhandled o2::framework::runtime_error reached the top of main of o2-sim-digitizer-workflow, device shutting down. Reason: Fatal error
[ERROR] pid 31139 (internal-dpl-ccdb-backend) crashed with or was killed with exit code 1
[ERROR]  - Device internal-dpl-ccdb-backend: pid 31139 (exit 1)
[INFO]    - First error: [15:45:50][FATAL] Unable to find object CTP/Calib/OrbitReset/1550600800000
[ERROR] SEVERE: Device internal-dpl-ccdb-backend (31139) had at least one message above severity 5: Unable to find object CTP/Calib/OrbitReset/1550600800000


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/a6917ed4e7431b2725430c07aae69728ba537c46/slc8_x86-64/o2checkcode/1.0-local30/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12079/c1e9bcaa8af2752225fb1fd1165e3715813d0243/build_O2_fullCI/pretty.html).

martenole commented:
Ok I think it would be good to test this on the FLP199. @shahor02 would it be possible to manually install an O2 version with these changes included on the FLP and run the TRD DCS DP processing with that for a few minutes? Leonardo has tested this with simulated data, but before merging I would like to try it on the real system. Maybe we could do it some time this week?

martenole commented:
Merging went wrong. This PR will be closed and re-opened with only the relevant (TRD) changes included
